woocommerce/grow
# Add `nvm use` and auto-branching

### Changes proposed in this Pull Request:

This PR extends `release` command with additional parameters
```
--nvm_use        - used to tell `release` to use `nvm use` per project basis.
--default_branch - used to pass project's default repo branch name into `release` command.
```
PR also adds few commands to be using as complementary commands to the overwritten `release` command:
```
simulate - used to run simulation mode on the new release command.
branch   - used to add a dialog for release branch creation.
```
Added set of tools for `Git`, `Nvm` and `Product`.

It is now possible to run the `release` and `simulate` commands using new parameters and see some changes in the command workflows.
```bash
./bin/wr simulate --nvm_use --default_branch develop --generate_changelog --wc_tested 8.3 --wp_tested 6.4 https://github.com/woocommerce/pinterest-for-woocommerce/tree/testrelease/1.3.13
```
For example if you provide a GitHub url using a branch name which does not exist on the repository, you will see a prompt to create it and release from there.

<img width="1551" alt="Terminal" src="https://github.com/woocommerce/grow/assets/9010963/1d1d5ee3-eb06-418d-8591-992a3ae66ef2">

It is also possible to skip release branch creation and release from the default which is set to `develop` by default. All you need is to refuse release branch creation.

<img width="1551" alt="Terminal" src="https://github.com/woocommerce/grow/assets/9010963/4cad2b22-6c56-463b-bb3f-9fea13ebe7ba">

The invisible part of changes introduced with this PR is `nvm use` command to run for every releasing product being inside the corresponding product directories to switch the release process to a proper `node` and `npm` versions. If `--nvm_use` flag is provided into `release` command, the additional logic will get involved into `build` process, if not - then the `release` will flow as usual.

### Detailed test instructions:


1. Checkout `add-nvm-and-autobranching` branch.
2. Open `release.txt` in the root and add some lines to it.
```txt
pinterest-for-woocommerce	1.3.13	testrelease/1.3.13
woocommerce-brands	1.6.62	testrelease/1.6.62
```
**Note:** I am using Pinterest and Brands here because both projects have to use their own `node` version and w/o `nvm use` the build will fail. Pinterest can build with `node 14` when Brands require `node 16` for release.

You can provide any nonsense instead of a release branch names to check how branch prompt is working.
3. Run bulk command
```bash
./bin/wr bulk simulate --generate_changelog --wc_tested 8.3 --wp_tested 6.4
```
**Note:** I am using `bulk` command here as a single command to work with `release.txt`.  Also you may notice `--wc_tested` and `--wp_tested` flags, those are also here to check version bump still functional.
4. While releasing wait for a branch creation prompt to appear.
5. Choosing `y` will start a branch creation process. Branch will be created from a default one and pushed to remote. The push is requires for a `changelog` command to generate its changelog part.
6. Choosing `N` will drop release branch creation process and prompt to release from a default branch which is, if rejected, will terminate the release process.
7. When running the simulation every release branch, except the default, must be cleaned up automatically at the end of the process.
8. All the steps are followed with the corresponding console messages/notifications/errors.
### Additional details:


### Changelog entry

> Dev - Adding auto-release branching.
> Dev - Adding `nvm use` on product basis.
